### PR TITLE
[JAX] Unifying GeLU and GeGLU in LayerNorm MLP

### PR DESCRIPTION
**Pull request by phu0ngng**
This PR unifies the GeLU and GEGLU implementations in `LayerNormMLP` via a generalized `fused_layernorm_fp8_mlp`. Previously, there were two separate APIs for the two mentioned activations. The new routine takes `activation_type: Tuple` and `use_bias: bool` as two additional arguments, compared to old routines.

This is a preparation step for adding more activations (i.e. `swiglu`) later.

### CODE REVIEW

File: transformer_engine/jax/mlp.py [hunk 1]
```diff
@@ -23,369 +23,56 @@
...
     """
-    Gated gelu
+    Activation Unit
     """
-    assert x.shape[-2] == 2    # Linear + GeLU
-
-    output = _geglu(x)
-
+    if 'linear' in activation_type:
+        assert x.shape[-2] == 2 # Linear + GeLU
```
Comment (zlsh80826) [inline transformer_engine/jax/mlp.py:46]: How about something like
```
if len(activation_type) > 1:
    assert x.shape[-2] == len(activation_type)
```
Not sure if there are more complicated gate/activation cases in the future

File: transformer_engine/jax/mlp.py [hunk 2]
```diff
@@ -23,369 +23,56 @@
...
-
-
-def _gelu_bwd_rule(ctx, g):
-    x, = ctx
-    assert x.dtype == g.dtype
-
-    dx = dgelu(g, x)
-    dx = jnp.reshape(dx, x.shape)
-    return (dx,)
+activation_lu_dict = {
```
Comment (zlsh80826) [inline transformer_engine/jax/mlp.py:26]: Nitpick: variable name `activation_dict` seems better

File: transformer_engine/jax/mlp.py [hunk 3]
```diff
@@ -470,13 +163,16 @@ def _layernorm_gelu_fp8_mlp_fwd_rule(
...
+        activation_type,
+        use_bias):
 
+    is_gated = 'linear' in activation_type
     # x should be in shape of (batch..., hidden)
     # Kernel_1 should be in shape of (Hidden_in, 1, Hidden_out)
     # Kernel_2 should be in shape of (Hidden_in, Hidden_out)
     assert len(kernel_1.shape) == 3
-    assert kernel_1.shape[-2] == 1
+    assert kernel_1.shape[-2] == 2 if is_gated else 1
```
Comment (zlsh80826) [inline transformer_engine/jax/mlp.py:175]:
```suggestion
    assert kernel_1.shape[-2] == len(activation_type)
```

File: tests/jax/test_custom_call_compute.py [hunk 4]
```diff
@@ -211,118 +227,10 @@ def _convert_to_activation_function(fn_or_string):
...
-                                          fp8_metas_scale_inv)
-            return jnp.mean(
-                layernorm_gelu_fp8_mlp(x, ln_s, None, [y, z], [w, v], fp8_meta_pkg, "rmsnorm"))
-
-        def ln_gelu_fp8_mlp_ref(x: jnp.ndarray, ln_scale: jnp.ndarray, kernel_1: jnp.ndarray,
+        def layernorm_fp8_mlp_ref(x: jnp.ndarray, ln_scale: jnp.ndarray, kernel_1: jnp.ndarray,
                                 kernel_2: jnp.ndarray, bias_1: jnp.ndarray, bias_2: jnp.ndarray,
                                 fp8_maxs: jnp.ndarray, amax: jnp.ndarray, scale: jnp.ndarray,
-                                scale_inv: jnp.ndarray) -> jnp.ndarray:
+                                  scale_inv: jnp.ndarray) -> jnp.ndarray:
```
Comment (ptrendx) [inline tests/jax/test_custom_call_compute.py:233]: nit: additional spaces

File: transformer_engine/jax/mlp.py [hunk 5]
```diff
@@ -633,36 +341,62 @@ def _layernorm_gelu_fp8_mlp_bwd_rule(
...
+        dactivation_lu_cast_transpose(
+            dgrad_2,
+            dot_1_output,
+            dactivation_lu_amax,
+            dactivation_lu_scale,
+            dactivation_lu_scale_inv,
+            bwd_dtype,
+            static_axis_boundary=-1)
+    else:
+        raise NotImplementedError
```
Comment (denera) [inline transformer_engine/jax/mlp.py:373]: @zlsh80826 @mingxu1067 @ptrendx Is there a reason why we can't support fused biases here via `dactivation() -> cast_transpose_dbias()`?

If we ultimately don't add partially fused biases into the underlying custom call, then we should probably have a useful error message here stating that unfused biases are supported for all activations via `te.flax.LayerNormMLP`.

**Review by denera — APPROVED** [review]
LGTM!

**Comment by denera** [comment]
@phu0ngng do you have CI permissions yet? If not, please check in with @ptrendx to get permissions and then trigger the CI run for this. We can merge once the tests come back clean.

**Comment by denera** [comment]
@phu0ngng Also please fix the linting errors here before the CI. Thanks!
```
************* Module transformer_engine.jax.flax.module
transformer_engine/jax/flax/module.py:690:26: W1401: Anomalous backslash in string: '\m'. String constant might be missing an r prefix. (anomalous-backslash-in-string)
transformer_engine/jax/flax/module.py:690:42: W1401: Anomalous backslash in string: '\s'. String constant might be missing an r prefix. (anomalous-backslash-in-string)
transformer_engine/jax/flax/module.py:690:48: W1401: Anomalous backslash in string: '\m'. String constant might be missing an r prefix. (anomalous-backslash-in-string)
transformer_engine/jax/flax/module.py:690:66: W1401: Anomalous backslash in string: '\e'. String constant might be missing an r prefix. (anomalous-backslash-in-string)
transformer_engine/jax/flax/module.py:691:17: W1401: Anomalous backslash in string: '\g'. String constant might be missing an r prefix. (anomalous-backslash-in-string)
transformer_engine/jax/flax/module.py:696:51: W1401: Anomalous backslash in string: '\g'. String constant might be missing an r prefix. (anomalous-backslash-in-string)
transformer_engine/jax/flax/module.py:702:64: W1401: Anomalous backslash in string: '\g'. String constant might be missing an r prefix. (anomalous-backslash-in-string)
```

**Review by zlsh80826 — APPROVED** [review]
LGTM

**Review by mingxu1067 — APPROVED** [review]
-LGTM-

**Review by mingxu1067 — CHANGES_REQUESTED** [review]
Kindly remove the unnessary comments

File: transformer_engine/jax/mlp.py (hunk 5 above)
Comment (denera) [reply transformer_engine/jax/mlp.py:373]: @mingxu1067 This is my ignorance with XLA's internal mechanisms but I'm surprised that it's more performant to do `LayerNormDenseGeneral w/ biases-> Activation -> DenseGeneral w/biases` and let JAX autodiff handle the backward pass than to implement a custom BWD with `dactivation() -> cast_transpose_dbias()`. Does XLA construct a better fused kernel in for this particular autodiff'd BWD than what we can do ourselves?

In any case, it should be pretty easy to template the `dactivation_dbias_cast_transpose()` kernel in the next PR to address this without running into any potential issues with separate `dactivation() -> cast_transpose_dbias()` calls.

Comment (mingxu1067) [reply transformer_engine/jax/mlp.py:373]: > Does XLA construct a better fused kernel in for this particular autodiff'd BWD than what we can do ourselves?
As I know, XLA does not do well on `cast_transpose` related fusion. IIRC, the perf is about 50% slower than TE's custom calls

File: tests/jax/test_custom_call_compute.py [hunk 6]
```diff
@@ -401,12 +321,15 @@ def ref_func(x, ln_s, y, z, w, v, fp8_max, fp8_metas_amax, fp8_metas_scale,
...
-                        dtype=jnp.bfloat16)
+        if use_bias:
+            assert_allclose(jnp.asarray(primitive_b1_grad, np.float32),
+                            jnp.asarray(ref_b1_grad, np.float32),
+                            dtype=jnp.bfloat16)
+            assert_allclose(jnp.asarray(primitive_b2_grad, np.float32),
+                            jnp.asarray(ref_b2_grad, np.float32),
+                            dtype=jnp.bfloat16)
+
+    #test_grad_fused_layernorm_fp8_mlp(256, 256, 256, activation_type = ('gelu', 'linear' ), use_bias = False)
```
Comment (denera) [inline tests/jax/test_custom_call_compute.py:332]:
```suggestion
```

@phu0ngng We can merge this PR once this is done.

**Review by denera — CHANGES_REQUESTED** [review]

**Comment by phu0ngng** [comment]
Hi @denera, @mingxu1067,
I resolved all of your change requests.
Please have a look and let me know if you have any other suggestions.

**Review by mingxu1067 — APPROVED** [review]
LGTM

**Review by denera — APPROVED** [review]

**Comment by zlsh80826** [comment]
Congratulations on your first pull request! @phu0ngng 
This really help the future maintainance for various activation types!

**Pull request merged**